### PR TITLE
!str: #16902: Unified stream internals

### PR DESCRIPTION
**Pull request by drewhk**

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala [hunk 1]
```diff
@@ -23,10 +25,35 @@ import akka.http.util._
...
+    override val outlets: Seq[Outlet[_]] = bytesOut :: httpResponses :: Nil
+
+    override def deepCopy(): Shape = HttpClientPorts(
+      new Inlet(bytesIn.toString),
+      new Outlet(bytesOut.toString),
+      new Inlet(httpResponses.toString),
+      new Outlet(httpRequests.toString))
+
+    override def copyFromPorts(inlets: Seq[Inlet[_]], outlets: Seq[Outlet[_]]): Shape = {
+      val bIn :: htpIn :: Nil = inlets
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/engine/client/HttpClient.scala]: is there a nicer way here?

File: akka-http-core/src/main/scala/akka/http/engine/rendering/RenderSupport.scala [hunk 2]
```diff
@@ -30,25 +29,24 @@ private object RenderSupport {
...
-      first.to(Sink(flowSubscriber)).run()(materializer)
-      second.to(Sink.cancelled).run()(materializer)
-    }
+  def CancelSecond[T](first: Source[T, _], second: Source[T, _]): Source[T, Unit] = {
+    Source(first) { implicit b ⇒
+      frst ⇒
+        import FlowGraph.Implicits._
+        second ~> Sink.cancelled
+        frst.outlet
+    }.mapMaterialized((_) ⇒ ())
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/engine/rendering/RenderSupport.scala]: this ensures unit as the result, but maybe it makes more sense here to propagate the first materialized value?

File: akka-http-core/src/main/scala/akka/http/engine/server/HttpServer.scala [hunk 3]
```diff
@@ -18,17 +20,37 @@ import akka.http.engine.parsing.ParserOutput._
...
+    override def deepCopy() = HttpServerPorts(
+      new Inlet(bytesIn.toString),
+      new Outlet(bytesOut.toString),
+      new Inlet(httpRequests.toString),
+      new Outlet(httpResponses.toString))
+
+    override def copyFromPorts(inlets: immutable.Seq[Inlet[_]], outlets: immutable.Seq[Outlet[_]]): Shape = {
+      require(inlets.size == 2, s"proposed inlets [${inlets.mkString(", ")}] do not fit BidiShape")
+      require(outlets.size == 2, s"proposed outlets [${outlets.mkString(", ")}] do not fit BidiShape")
+      HttpServerPorts(inlets(0).asInstanceOf[Inlet[ByteString]], outlets(0).asInstanceOf[Outlet[ByteString]],
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/engine/server/HttpServer.scala]: indexing into list? well, it is a small list, so may be fine.

File: akka-http-core/src/main/scala/akka/http/engine/server/HttpServer.scala [hunk 4]
```diff
@@ -89,39 +108,44 @@ private[http] object HttpServer {
...
-    import FlowGraphImplicits._
+        val bypassFanout = b.add(Broadcast[RequestOutput](2, OperationAttributes.name("bypassFanout")))
+        val bypassMerge = b.add(new BypassMerge(settings, log))
+        val bypassInput = bypassMerge.in0
+        val bypassOneHundredContinueInput = bypassMerge.in1
+        val bypassApplicationInput = bypassMerge.in2
 
-    Flow() { implicit b ⇒
-      //FIXME: the graph is unnecessary after fixing #15957
-      transportIn ~> requestParsing ~> bypassFanout ~> requestPreparation ~> serverFlow ~> bypassMerge.applicationInput ~> rendererPipeline ~> transportOut
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/engine/server/HttpServer.scala]: this previous wiring was maybe nicer

File: akka-http-core/src/main/scala/akka/http/model/HttpEntity.scala [hunk 5]
```diff
@@ -100,13 +100,13 @@ sealed trait BodyPartEntity extends HttpEntity with japi.BodyPartEntity {
 sealed trait RequestEntity extends HttpEntity with japi.RequestEntity with ResponseEntity {
   def withContentType(contentType: ContentType): RequestEntity
 
-  override def transformDataBytes(transformer: Flow[ByteString, ByteString]): RequestEntity
+  override def transformDataBytes(transformer: Flow[ByteString, ByteString, _]): RequestEntity
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/model/HttpEntity.scala]: Unit or _ makes more sense?

File: akka-http-core/src/main/scala/akka/http/model/HttpEntity.scala [hunk 6]
```diff
@@ -38,13 +38,13 @@ sealed trait HttpEntity extends japi.HttpEntity {
...
    */
-  def dataBytes: Source[ByteString]
+  def dataBytes: Source[ByteString, Unit]
 
   /**
    * Collects all possible parts and returns a potentially future Strict entity for easier processing.
    * The Future is failed with an TimeoutException if the stream isn't completed after the given timeout.
    */
-  def toStrict(timeout: FiniteDuration)(implicit fm: FlowMaterializer): Future[HttpEntity.Strict] = {
+  def toStrict(timeout: FiniteDuration)(implicit fm: ActorFlowMaterializer): Future[HttpEntity.Strict] = {
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/model/HttpEntity.scala]: changed to ActorFlowMaterializer due to my rename. I am not sure where is enough to have a FlowMaterializer, but we should prefer it whenever possible
.

File: akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala [hunk 7]
```diff
@@ -94,9 +93,10 @@ private[http] object StreamUtils {
     Flow[ByteString].section(name("sliceBytes"))(_.transform(() ⇒ transformer))
   }
 
-  def limitByteChunksStage(maxBytesPerChunk: Int): Stage[ByteString, ByteString] =
+  def limitByteChunksStage(maxBytesPerChunk: Int): PushPullStage[ByteString, ByteString] =
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala]: this change is no longer necessary

File: akka-stream-testkit/src/test/scala/akka/stream/testkit/ChainSetup.scala [hunk 8]
```diff
@@ -7,20 +7,23 @@ import org.reactivestreams.Publisher
 import akka.stream.ActorFlowMaterializer
 
 class ChainSetup[In, Out](
-  stream: Flow[In, In] ⇒ Flow[In, Out],
+  stream: Flow[In, In, _] ⇒ Flow[In, Out, _],
```
Comment (drewhk) [inline akka-stream-testkit/src/test/scala/akka/stream/testkit/ChainSetup.scala]: this style is remnant of the old Flow where it was what now Source is. We should refactor these to use via. Probably not in this PR but we should open a ticket.

File: akka-stream-tests/src/test/java/akka/stream/javadsl/FlowTest.java [hunk 9]
```diff
@@ -297,51 +330,14 @@ public void apply(Pair<String, Integer> param) throws Exception {
   }
 
   @Test
-  public void mustBeAbleToUseUnzip() {
```
Comment (drewhk) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/FlowTest.java]: this should be revived?

File: akka-stream-tests/src/test/scala/akka/stream/DslConsistencySpec.scala [hunk 10]
```diff
@@ -63,14 +53,12 @@ class DslConsistencySpec extends WordSpec with Matchers {
...
-      ("Key" -> List(sKeyClass, jKeyClass)) ::
-      ("MaterializedMap" -> List(sMaterializedMapClass, jMaterializedMapClass)) ::
-      ("FlowGraph" -> List(sFlowGraphClass, jFlowGraphClass)) ::
-      ("PartialFlowGraph" -> List(sPartialFlowGraphClass, jPartialFlowGraphClass)) ::
+      ("RunanbleFlow" -> List(sRunnableFlowClass, jRunnableFlowClass)) ::
       Nil foreach {
         case (element, classes) ⇒
 
           s"provide same $element transforming operators" in {
+            pending
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/DslConsistencySpec.scala]: wanna fix now? or open a ticket?

File: akka-stream-tests/src/test/scala/akka/stream/DslFactoriesConsistencySpec.scala [hunk 11]
```diff
@@ -35,34 +35,26 @@ class DslFactoriesConsistencySpec extends WordSpec with Matchers {
...
-
-  val sFlow = classOf[scaladsl.Flow[_, _]]
-  val jFlow = classOf[javadsl.Flow[_, _]]
+  val sFlow = classOf[scaladsl.Flow[_, _, _]]
+  val jFlow = classOf[javadsl.Flow[_, _, _]]
 
   "Java DSL" must provide {
     "Source" which {
       "allows creating the same Sources as Scala DSL" in {
+        pending
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/DslFactoriesConsistencySpec.scala]: fix now or open a ticket?

File: akka-stream-tests/src/test/scala/akka/stream/actor/ActorPublisherSpec.scala [hunk 12]
```diff
@@ -262,12 +256,9 @@ class ActorPublisherSpec extends AkkaSpec with ImplicitSender {
...
       val sink = Sink[String](receiverProps(probe.ref))
 
-      val mat = source.collect {
+      val (snd, rcv) = source.collect {
         case n if n % 2 == 0 ⇒ "elem-" + n
-      }.to(sink).run()
-
-      val snd = mat.get(source)
-      val rcv = mat.get(sink)
+      }.toMat(sink)(Keep.both).run()
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/actor/ActorPublisherSpec.scala]: Ah, there is a Keep.both. I just used Pair.apply

File: akka-stream-tests/src/test/scala/akka/stream/impl/InterpreterSpec.scala [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+package akka.stream.impl
+
+import akka.stream.Supervision._
+import akka.stream.testkit.AkkaSpec
+import akka.stream._
+import akka.stream.scaladsl._
+import akka.stream.testkit.StreamTestKit._
+import akka.stream.impl.fusing.ActorInterpreter
+
+class InterpreterSpec extends AkkaSpec {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/InterpreterSpec.scala]: Why does this file exist? Shouldn't it be part of the fusing package along with other tests?

File: akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+      runnable0123a.inPorts.size should be(0)
+      runnable0123a.outPorts.size should be(0)
+      runnable0123a.isRunnable should be(true)
+      runnable0123a.isFlow should be(false)
+      runnable0123a.isSink should be(false)
+      runnable0123a.isSource should be(false)
+    }
+
+    "be able to model hierarchic linear modules" in {
+      pending
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/StreamLayoutSpec.scala]: I won't fix these now, but we need more tests. It is a huge benefit of the new materialization system that the logic can be tested in isolation. We should use this power. (also, this is one lession from remoting)

File: akka-stream-tests/src/test/scala/akka/stream/io/StreamTcpSpec.scala [hunk 15]
```diff
@@ -375,48 +390,42 @@ class StreamTcpSpec extends AkkaSpec with TcpHelper {
...
+      val binding3F = bind.to(Sink(probe3)).run()
       probe3.expectErrorOrSubscriptionFollowedByError()
 
-      // The unbind should NOT fail even though the bind failed.
-      Await.result(binding.unbind(mm2), 1.second)
-      Await.result(binding.unbind(mm3), 1.second)
-
-      an[BindFailedException] shouldBe thrownBy { Await.result(binding.localAddress(mm2), 1.second) }
-      an[BindFailedException] shouldBe thrownBy { Await.result(binding.localAddress(mm3), 1.second) }
+      an[BindFailedException] shouldBe thrownBy { Await.result(binding2F, 1.second) }
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/io/StreamTcpSpec.scala]: unbinding changed now in that it makes only sense once you got a ServerBinding. So these semantic changes are valid IMHO

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala [hunk 16]
```diff
@@ -178,109 +174,35 @@ class FlowGraphCompileSpec extends AkkaSpec {
       }.run()
     }
 
-    "attachSource and attachSink" in {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala]: I believe these removed tests are no longer relevant but correct me if you think otherwise.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala [hunk 17]
```diff
@@ -290,240 +212,84 @@ class FlowGraphCompileSpec extends AkkaSpec {
...
+        import FlowGraph.Implicits._
+        b.add(in1 via f1) ~> b.add(out1)
       }.run()
-      FlowGraph { implicit b ⇒
-        import FlowGraphImplicits._
-        in1.via(f1) ~> f2.to(out1)
-      }.run()
-    }
-
-    "build all combinations with implicits" when {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphCompileSpec.scala]: Same here, I am not sure they are valuable anymore. We can revive them though.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphInitSpec.scala [hunk 18]
```diff
@@ -1,125 +0,0 @@
...
-            log.error(e, "FlowGraph init failure")
-            None
-        }
-      }
-
-      val graphs = Vector.fill(5)(Future(create()))
-      val result = Await.result(Future.sequence(graphs), 5.seconds).flatten.size should be(5)
-    }
-
-    "fail when the same `KeyedSink` is used in it multiple times" in {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGraphInitSpec.scala]: However, these are really irrelevant now.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSectionSpec.scala [hunk 19]
```diff
@@ -19,23 +23,37 @@ class FlowSectionSpec extends AkkaSpec(FlowSectionSpec.config) {
...
       Source.single(1).via(flow).to(Sink.ignore).run()
 
-      receiveN(1).foreach {
-        case s: String ⇒ s should include("my-dispatcher")
-      }
+      expectMsgType[String] should include("my-dispatcher1")
+    }
+
+    "have multiple levels of nesting" in {
+      val flow = Flow[Int].section(dispatcher("my-dispatcher1"))(
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSectionSpec.scala]: we can remove section actually. Calling withAttributes on a Flow, then using simple via can work as an alternative.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala [hunk 20]
```diff
@@ -48,71 +58,22 @@ object FlowSpec {
...
-        case o                   ⇒ ActorProcessorFactory.props(this, o)
-      }
-      val impl = actorOf(props.withDispatcher(settings.dispatcher), s"$flowName-$n-${op.attributes.name}")
-      (ActorProcessorFactory(impl), MaterializedMap.empty)
-    }
-
+  val faultyFlow: Flow[Any, Any, _] ⇒ Flow[Any, Any, _] = in ⇒ in.andThenMat { () ⇒
+    val props = Props(new BrokenActorInterpreter(settings, List(fusing.Map({ x: Any ⇒ x }, stoppingDecider)), "a3"))
+      .withDispatcher("akka.test.stream-dispatcher")
+    val processor = ActorProcessorFactory[Any, Any](system.actorOf(
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSpec.scala]: yes, I am attaching the broken processor to the system. we can probably fix this.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSupervisionSpec.scala [hunk 21]
```diff
@@ -20,11 +20,10 @@ class FlowSupervisionSpec extends AkkaSpec {
 
   val exc = new RuntimeException("simulated exc") with NoStackTrace
 
-  val failingMap = (s: Source[Int]) ⇒ s.map(n ⇒ if (n == 3) throw exc else n)
+  val failingMap = Flow[Int].map(n ⇒ if (n == 3) throw exc else n)
 
-  // FIXME this would be more elegant with Flow[Int, Int] and `via`, but `via` is currently not propagating the OperationAttributes
-  def run(s: Source[Int] ⇒ Source[Int]): immutable.Seq[Int] =
-    Await.result(s(Source(1 to 5)).grouped(1000).runWith(Sink.head), 3.seconds)
+  def run(f: Flow[Int, Int, Unit]): immutable.Seq[Int] =
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSupervisionSpec.scala]: another fixme down :)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSupervisionSpec.scala [hunk 22]
```diff
@@ -35,8 +34,7 @@ class FlowSupervisionSpec extends AkkaSpec {
     }
 
     "support resume " in {
-      val result = run(s ⇒ s.section(supervisionStrategy(Supervision.resumingDecider))(
-        failingMap(_)))
+      val result = run(failingMap.withAttributes(supervisionStrategy(Supervision.resumingDecider)))
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowSupervisionSpec.scala]: ... because, the new and almighty withAttributes is here to help us :)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala [hunk 23]
```diff
@@ -8,32 +8,38 @@ import akka.stream.ActorFlowMaterializerSettings
...
-    inMerge ~> Flow[Int].map(_ * 2) ~> m2 ~> Flow[Int].map(_ / 2).map(i ⇒ (i + 1).toString) ~> outMerge
-    source2 ~> inMerge
-    source3 ~> m2
-    source4 ~> outMerge
+    val inMerge = b.add(Merge[Int](2))
+    val outMerge = b.add(Merge[String](2))
+    val m2 = b.add(Merge[Int](2))
+
+    inMerge.out.map(_ * 2) ~> m2.in(0)
+    m2.out.map(_ / 2).map(i ⇒ (i + 1).toString) ~> outMerge.in(0)
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBackedFlowSpec.scala]: oh boy, how I love this notation

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphConcatSpec.scala [hunk 24]
```diff
@@ -97,6 +104,7 @@ class GraphConcatSpec extends TwoStreamsSetup {
     }
 
     "work with one delayed failed and first nonempty publisher" in {
+      pending // FIXME: This relies on materialization order!!
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphConcatSpec.scala]: this worked before because of the particular order we materialized things. We should revisit this later.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphOpsIntegrationSpec.scala [hunk 25]
```diff
@@ -1,59 +1,51 @@
...
-import akka.stream.scaladsl.FlowGraphImplicits._
 import akka.stream.testkit.AkkaSpec
 import akka.stream.testkit.StreamTestKit.{ OnNext, SubscriberProbe }
 import akka.util.ByteString
+import akka.stream.{ Inlet, Outlet, Shape, Graph }
 
 object GraphOpsIntegrationSpec {
+  import FlowGraph.Implicits._
 
-  object Lego {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphOpsIntegrationSpec.scala]: Changed to the Shuffler below to have two in and out ports instead of one.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/OptimizingActorBasedFlowMaterializerSpec.scala [hunk 26]
```diff
@@ -1,52 +0,0 @@
...
-
-import akka.stream.scaladsl.OperationAttributes._
-import akka.stream.{ ActorFlowMaterializer, ActorFlowMaterializerSettings, Optimizations }
-import akka.stream.testkit.AkkaSpec
-import akka.testkit._
-
-import scala.concurrent.duration._
-import scala.concurrent.Await
-
-class OptimizingActorFlowMaterializerSpec extends AkkaSpec with ImplicitSender {
```
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/OptimizingActorBasedFlowMaterializerSpec.scala]: Sadly, this is gone temporarily.

File: akka-stream/src/main/resources/reference.conf [hunk 27]
```diff
@@ -34,10 +34,6 @@ akka {
         # for cancelation (see `akka.stream.subscription-timeout.mode`)
         timeout = 5s
       }
-
-      # Fully qualified config path which holds the dispatcher configuration
-      # to be used by FlowMaterialiser when creating Actors for IO operations.
-      file-io-dispatcher = ${akka.io.tcp.file-io-dispatcher}
```
Comment (drewhk) [inline akka-stream/src/main/resources/reference.conf]: this is migrated to akka-http where it belongs

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala [hunk 28]
```diff
@@ -35,11 +31,11 @@ object ActorFlowMaterializer {
    * the processing steps. The default `namePrefix` is `"flow"`. The actor names are built up of
    * `namePrefix-flowNumber-flowStepNumber-stepName`.
    */
-  def apply(materializerSettings: Option[ActorFlowMaterializerSettings] = None, namePrefix: Option[String] = None)(implicit context: ActorRefFactory): ActorFlowMaterializer = {
+  def apply(materializerSettings: Option[ActorFlowMaterializerSettings] = None, namePrefix: Option[String] = None, optimizations: Optimizations = Optimizations.none)(implicit context: ActorRefFactory): ActorFlowMaterializer = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: optimizations are ignored for now. In fact I am no longer sure that the optimizer should be here. It is much better to move this to a separate optimizer so you don't need to reoptimize every time the same stuff.

File: akka-stream/src/main/scala/akka/stream/extra/Implicits.scala [hunk 29]
```diff
@@ -18,19 +18,19 @@ object Implicits {
...
    */
-  implicit class TimedSourceDsl[I](val flow: Source[I]) extends AnyVal {
+  implicit class TimedSourceDsl[I, Mat](val source: Source[I, Mat]) extends AnyVal {
 
     /**
      * Measures time from receieving the first element and completion events - one for each subscriber of this `Flow`.
      */
-    def timed[O](measuredOps: Source[I] ⇒ Source[O], onComplete: FiniteDuration ⇒ Unit): Source[O] =
-      Timed.timed[I, O](flow, measuredOps, onComplete)
+    def timed[O, Mat2](measuredOps: Source[I, Mat] ⇒ Source[O, Mat2], onComplete: FiniteDuration ⇒ Unit): Source[O, Mat2] =
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/extra/Implicits.scala]: another case of Source => Source pattern which just remained from the time when we only had Flow which was actually a source.

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala [hunk 30]
```diff
@@ -4,449 +4,150 @@
...
-          case (f: Ast.Fused) :: _ ⇒ throw new IllegalStateException("Fused AST nodes not allowed to be present in the input to the optimizer: " + f)
-          //TODO Ast.Take(-Long.MaxValue..0) == stream doesn't do anything. Perhaps output warning for that?
-          case noMatch             ⇒ noMatch
-        }
+  override def materialize[Mat](runnableFlow: RunnableFlow[Mat]): Mat = {
+    runnableFlow.module.validate()
 
-      // The `elide` phase
-      // TODO / FIXME : This phase could be pulled out to be executed incrementally when building the Ast
-      def elide(rest: List[Ast.AstNode], orig: List[Ast.AstNode]): List[Ast.AstNode] =
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: sorry, temporarily gone :(

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala [hunk 31]
```diff
@@ -550,41 +206,39 @@ private[akka] class StreamSupervisor(settings: ActorFlowMaterializerSettings) ex
  * INTERNAL API
  */
 private[akka] object ActorProcessorFactory {
+  import akka.stream.impl.Stages._
 
-  import Ast._
-  def props(materializer: ActorFlowMaterializer, op: AstNode): Props = {
-    val settings = materializer.settings // USE THIS TO AVOID CLOSING OVER THE MATERIALIZER BELOW
+  def props(materializer: ActorFlowMaterializerImpl, op: StageModule, parentAttributes: OperationAttributes): (Props, Any) = {
+    val att = parentAttributes and op.attributes
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: this line and the one below is an actual fix, and the problem was exposed by the supervision additions.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: nice to see hiearchic attributes working in practice though

File: akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala [hunk 32]
```diff
@@ -0,0 +1,301 @@
...
+     */
+    def graph[S <: Shape](graph: Graph[S, _]): S = delegate.add(graph)
+
+    def source[T](source: Source[T, _]): Outlet[T] = delegate.add(source.asScala)
+
+    def sink[T](sink: Sink[T, _]): Inlet[T] = delegate.add(sink.asScala)
+
+    def run(mat: ActorFlowMaterializer): Unit = delegate.buildRunnable().run()(mat)
+  }
+}
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala]: 300 lines less :)

File: akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala [hunk 33]
```diff
@@ -7,19 +7,21 @@ import akka.actor.ActorRef
...
-
 import scala.concurrent.Future
+import akka.stream.impl.StreamLayout
 
 /** Java API */
 object Sink {
 
   import akka.stream.scaladsl.JavaConverters._
 
+  val factory: SinkCreate = new SinkCreate {}
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala]: what is this?

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala [hunk 34]
```diff
@@ -16,21 +17,24 @@ import scala.concurrent.duration.FiniteDuration
...
 import scala.language.implicitConversions
 import akka.stream.stage.Stage
+import akka.stream.impl.StreamLayout
 
 /** Java API */
 object Source {
 
   import scaladsl.JavaConverters._
 
+  val factory: SourceCreate = new SourceCreate {}
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: what is this?

File: akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala (hunk 32 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala]: sorry, 1100

File: akka-stream/src/main/scala/akka/stream/scaladsl/FlowGraph.scala [hunk 35]
```diff
@@ -1,1491 +0,0 @@
...
-      builder.addEdge(source, flow, junctionIn)
-      junctionIn.next
-    }
-
-    def ~>(sink: UndefinedSink[Out]): Unit =
-      builder.addEdge(source, flow, sink)
-
-    def ~>(sink: Sink[Out]): Unit =
-      builder.addEdge(source, flow, sink)
-  }
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/FlowGraph.scala]: *phew* nice to see this beast gone

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 36]
```diff
@@ -0,0 +1,261 @@
...
+
+    implicit def fanOut2flow[I, O](j: UniformFanOutShape[I, O])(implicit b: Builder): PortOps[O, Unit] =
+      new PortOps(find(b, j, 0), b)
+
+    implicit class SourceArrow[T](val s: Source[T, _]) extends AnyVal with CombinerBase[T] {
+      override def importAndGetPort(b: Builder): Outlet[T] = b.add(s)
+    }
+
+  }
+
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: 260 vs 1490 lines :)

File: akka-stream/src/main/scala/akka/stream/scaladsl/StreamTcp.scala [hunk 37]
```diff
@@ -32,46 +33,19 @@ import akka.stream.impl.io.StreamTcpManager
...
-    def connections: Source[IncomingConnection]
-
-    /**
-     * Asynchronously triggers the unbinding of the port that was bound by the materialization of the `connections`
-     * [[Source]] whose [[MaterializedMap]] is passed as parameter.
-     *
-     * The produced [[scala.concurrent.Future]] is fulfilled when the unbinding has been completed.
-     */
-    def unbind(materializedMap: MaterializedMap): Future[Unit]
+  case class ServerBinding(localAddress: InetSocketAddress)(private val unbindAction: () ⇒ Future[Unit]) {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/StreamTcp.scala]: nice old case classes now

**Comment by drewhk** [comment]
First round of self-review done

File: akka-docs-dev/rst/scala/code/docs/stream/GraphCyclesSpec.scala [hunk 38]
```diff
@@ -13,89 +13,90 @@ class GraphCyclesSpec extends AkkaSpec {
...
 
-        val merge = Merge[Int]
-        val bcast = Broadcast[Int]
+        val merge = b.add(Merge[Int](2))
+        val bcast = b.add(Broadcast[Int](2))
 
-        source ~> merge ~> Flow[Int].map { (s) => println(s); s } ~> bcast ~> Sink.ignore
-        bcast ~> merge
+        source ~> merge ~> Flow[Int].map { s => println(s); s } ~> bcast ~> Sink.ignore()
+                  merge                    <~                      bcast
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/stream/GraphCyclesSpec.scala]: neat ;]

File: akka-stream-tests/src/test/scala/akka/stream/actor/ActorPublisherSpec.scala (hunk 12 above)
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/actor/ActorPublisherSpec.scala]: `Pair.apply _` will allocate an instance for every use, `Keep.both` is preallocated

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala (hunk 28 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: absolutely; please file a ticket

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala [hunk 39]
```diff
@@ -272,11 +261,14 @@ final case class ActorFlowMaterializerSettings(
   def withOptimizations(optimizations: Optimizations): ActorFlowMaterializerSettings =
     copy(optimizations = optimizations)
 
-  private def isPowerOfTwo(n: Integer): Boolean = (n & (n - 1)) == 0 // FIXME this considers 0 a power of 2
+  private def requirePowerOfTwo(n: Integer, name: String): Unit = {
+    require(maxInputBufferSize > 0, s"$name must be > 0")
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: `s/maxInputBufferSize/n/`

**Comment by rkuhn** [comment]
> Test PASSed.
> 

YEEEEHAAAAAH!

File: akka-stream/src/main/scala/akka/stream/javadsl/Source.scala (hunk 34 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/Source.scala]: I encountered trouble with missing static forwarders for Java when just mixing in the traits, so I decided to go the safe route, which means offering the methods on an instance.

File: akka-stream-tests/src/test/scala/akka/stream/impl/InterpreterSpec.scala (hunk 13 above)
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/impl/InterpreterSpec.scala]: This is not related to fusing at all; it should probably be called ActorInterpreterSpec.

File: akka-stream-tests/src/test/java/akka/stream/javadsl/FlowTest.java (hunk 9 above)
Comment (rkuhn) [inline akka-stream-tests/src/test/java/akka/stream/javadsl/FlowTest.java]: no, the exact same test in in FlowGraphTest

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala (hunk 39 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: ah, thanks
